vpr/src/place/initial_placement.cpp

```diff
@@ -236,6 +257,191 @@ static bool is_block_placed(ClusterBlockId blk_id) {
     return (!(place_ctx.block_locs[blk_id].loc.x == INVALID_X));
 }
 
+static bool is_loc_legal(t_pl_loc loc, PartitionRegion& pr, t_logical_block_type_ptr block_type) {
+    const auto& grid = g_vpr_ctx.device().grid;
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+    bool legal = false;
+
+    //Check if the location is within its constraint region
+    for (auto reg : pr.get_partition_region()) {
+        if (reg.get_region_rect().contains(vtr::Point<int>(loc.x, loc.y))) {
+            //Check if blk_id type and the location physical types match
+            if (grid[loc.x][loc.y].type->index == block_type->index) {
+                //Check if the location has enough capacity left to place blk_id
+                if (place_ctx.grid_blocks[loc.x][loc.y].usage < grid[loc.x][loc.y].type->capacity) {
```

Comment:
I checked and found that we should check the capacity after determining the sub tiles, and it already exists in "macro_can_be_placed" function in place_util.cpp. So, I removed the condition because I don't need to check it beforehand.